src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
```

Comment:
Why not get rid of `Unexpected` and force users to always provide some string for `Expected`? 

Perhaps this could also serve for some inspiration: https://kowainik.github.io/posts/2018-09-25-co-log

Reply:
Yes, I think `Unexpected` is dead code rn, forgot to remove after adding `Expected`. I'll also gove the blog a read, thanks